presto-main/src/main/java/com/facebook/presto/type/setdigest/SetDigestFunctions.java

```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.type.setdigest;
+
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Throwables;
+import io.airlift.json.ObjectMapperProvider;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import static com.facebook.presto.type.setdigest.SetDigest.exactIntersectionCardinality;
+
+public final class SetDigestFunctions
+{
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    private SetDigestFunctions()
+    {
+    }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long cardinality(@SqlType(SetDigestType.NAME) Slice digest)
+    {
+        return SetDigest.newInstance(digest).cardinality();
+    }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long intersectionCardinality(@SqlType(SetDigestType.NAME) Slice slice1, @SqlType(SetDigestType.NAME) Slice slice2)
+    {
+        SetDigest digest1 = SetDigest.newInstance(slice1);
+        SetDigest digest2 = SetDigest.newInstance(slice2);
+
+        if (digest1.isExact() && digest2.isExact()) {
+            return exactIntersectionCardinality(digest1, digest2);
+        }
+
+        long cardinality1 = digest1.cardinality();
+        long cardinality2 = digest2.cardinality();
+        double jaccard = SetDigest.jaccardIndex(digest1, digest2);
+        digest1.mergeWith(digest2);
+        long result = Math.round(jaccard * digest1.cardinality());
+
+        // When one of the sets is much smaller than the other and approaches being a true
+        // subset of the other, the computed cardinality may exceed the cardinality estimate
+        // of the smaller set. When this happens the cardinality of the smaller set is obviously
+        // a better estimate of the one computed with the Jaccard Index.
+        return Math.min(result, Math.min(cardinality1, cardinality2));
+    }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double jaccardIndex(@SqlType(SetDigestType.NAME) Slice slice1, @SqlType(SetDigestType.NAME) Slice slice2)
+    {
+        SetDigest digest1 = SetDigest.newInstance(slice1);
+        SetDigest digest2 = SetDigest.newInstance(slice2);
+
+        return SetDigest.jaccardIndex(digest1, digest2);
+    }
+
+    @ScalarFunction
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice hashCounts(@SqlType(SetDigestType.NAME) Slice slice)
+    {
+        SetDigest digest = SetDigest.newInstance(slice);
+
+        try {
+            return Slices.utf8Slice(OBJECT_MAPPER.writeValueAsString(digest.getHashCounts()));
```

Comment:
What is hash counts?  If it is structured, shouldn't we return a row or map instead of JSON in a VARCHAR?  At the very least this should be return type JSON.

Reply:
The snippet will be similar to this one:
```
@ScalarFunction
@SqlType("map(bigint,smallint)")
public static Block hashCounts(@TypeParameter("map<bigint,smallint>") Type mapType, @SqlType(SetDigestType.NAME) Slice slice)
{
    SetDigest digest = SetDigest.newInstance(slice);

    // Maybe use static BlockBuilderStatus in order avoid `new`?
    BlockBuilder blockBuilder = mapType.createBlockBuilder(new BlockBuilderStatus(), 1);
    BlockBuilder singleMapBlockBuilder = blockBuilder.beginBlockEntry();
    for (Map.Entry<String, Short> entry : digest.getHashCounts().entrySet()) {
        VARCHAR.writeSlice(singleMapBlockBuilder, entry.getKey());
        SMALLINT.writeLong(singleMapBlockBuilder, entry.getValue());
    }
    blockBuilder.closeEntry();
    pageBuilder.declarePosition();

    return (Block) mapType.getObject(blockBuilder, 0);
}
```